src/directory_stack.rs

```diff
@@ -24,53 +24,86 @@ impl DirectoryStack {
-            println!("Directory stack is empty");
-            return;
-        }
-        if let Some(dir) = self.dirs.get(self.dirs.len() - 2) {
+        if let Some(dir) = self.get_previous_dir() {
             if let Err(err) = set_current_dir(dir) {
                 println!("{}: Failed to switch to directory {}", err, dir.display());
                 return;
             }
+        } else {
+            println!("Directory stack is empty");
+            return;
         }
-        self.dirs.pop_back();
+        self.dirs.pop_front();
         self.print_dirs();
     }
 
     pub fn pushd<I: IntoIterator>(&mut self, args: I)
         where I::Item: AsRef<str>
     {
-        self.change_and_push_dir(args);
-        self.print_dirs();
+        if let Some(dir) = args.into_iter().nth(1) {
+            self.change_and_push_dir(dir.as_ref());
+            self.print_dirs();
+        } else {
+            println!("No directory provided");
+            return;
+        }
     }
 
     pub fn cd<I: IntoIterator>(&mut self, args: I)
         where I::Item: AsRef<str>
     {
-        self.change_and_push_dir(args);
+        if let Some(dir) = args.into_iter().nth(1) {
+            let dir = dir.as_ref();
+            if dir == "-" {
+                self.switch_to_previous_directory();
+            } else {
+                self.change_and_push_dir(dir);
+            }
+        } else {
+            self.switch_to_home_directory();
+        }
     }
 
-    // TODO the signature for this function doesn't make a lot of sense I did
-    // it this way to for ease of use where it is used, however, it should take
-    // just one dir instead of args once we add features like `cd -`.
-    pub fn change_and_push_dir<I: IntoIterator>(&mut self, args: I)
-        where I::Item: AsRef<str>
-    {
-        if let Some(dir) = args.into_iter().skip(1).next() {
-            match (set_current_dir(dir.as_ref()), current_dir()) {
-                (Ok(()), Ok(cur_dir)) => {
-                    self.push_dir(cur_dir);
-                }
-                (Err(err), _) => {
-                    println!("Failed to set current dir to {}: {}", dir.as_ref(), err);
-                    return;
-                }
-                (_, _) => (),
+    fn switch_to_home_directory(&mut self) {
+        if let Some(home) = home_dir() {
+            if let Some(home) = home.to_str() {
+                self.change_and_push_dir(home);
+            } else {
+                println!("Failed to convert home directory to str");
             }
         } else {
-            println!("No directory provided");
-            return;
+            println!("Failed to get home directory");
+        }
+    }
+
+    fn switch_to_previous_directory(&mut self) {
+        if let Some(prev) = self.get_previous_dir()
+                                .map(|path| path.to_string_lossy().to_string()) {
+            println!("{}", prev);
+            self.change_and_push_dir(&prev);
```

Comment:
Actually, this is not the POSIX behavior. What POSIX does is to swap the last and next to last directories in the dirstack. I believe this is a better behavior, since you can already `popd`, anyways. POSIX's `cd -` serves a different purpose than `popd` (both are useful!), and I don't think duplicate functionality is really necessary in this case.